### DISCUSSION TITLE
Using Python class instead of modelx

### DISCUSSION THREAD

**Discussion by fumitoh** [fumitoh/modelx#75] [Q&A]
(Originally posted on https://modelx.io/blog/2022/10/29/why-you-should-use-modelx/)

There is a library called methodtools which includes an lru_cache decorator, someone has even opened an issue and proposed a solution for clearing the cache when an attribute is changed - [https://github.com/youknowo...](https://disq.us/url?url=https%3A%2F%2Fgithub.com%2Fyouknowone%2Fmethodtools%2Fissues%2F16%3AqvHvOvKCIWZE5REGKfUIC1CdAtw&cuid=6465175). This might solve the issue brought up under the section "Using Python class instead of modelx"?

**Comment by fumitoh** [comment]
This may work, but you still can't get rid of `self.`.
And I'm not sure how to update methods with this approach...

With modelx, you can do

```python
space = mx.new_space()

@mx.defcells
def foo():
    return bar()

@mx.defcells
def bar():
    return 1

space.foo()  # or just foo() works
# Returns 1

# Now updating bar
@mx.defcells
def bar():
    return 2

space.foo()  # or just foo() works
# Returns 2
```

**Reply by MatthewCaseres** [reply]
I think the user would update the definition in the class and then create a new instance?

**Reply by declann** [reply]
Sorry to dig up an old discussion but adding some thoughts here:

A thing I'm appreciating a lot in modelx vs classes is that it can be interactively updated both using standard Python tools like popular notebook tools and the more specialized Spyder-modelx environment. For development and exploration I think this is really neat.

An alternative classes approach to update methods is a new class inheriting from the existing class with a `bar` override - you can do this interactively. For cognitive load and keeping track I think modelx's approach adds a lot over classes.

edit: When a model is complete/in production - where interactively updating a model may be bad without extreme care, modelx and classes are closer (but some infrastructure around modelx is still really good)
